pdf_text_overlay/pdfWriter.py

```diff
@@ -11,152 +18,150 @@
-                    y0 = config_data['draw_shape']['y0-coordinate']
-                    y1 = config_data['draw_shape']['y1-coordinate']
+                # Get the cartesian coordinates
+                x0 = config_data['draw_shape'].get('x0-coordinate')
+                x1 = config_data['draw_shape'].get('x1-coordinate')
+                y0 = config_data['draw_shape'].get('y0-coordinate')
+                y1 = config_data['draw_shape'].get('y1-coordinate')
 
-                    can.line(x0*inch,y0*inch, x1*inch, y1*inch)
-                if config_data['draw_shape']['shape'] == 'Rectangle':
-                    x0 = config_data['draw_shape']['x0-coordinate']
-                    x1 = config_data['draw_shape']['x1-coordinate']
-                    y0 = config_data['draw_shape']['y0-coordinate']
-                    y1 = config_data['draw_shape']['y1-coordinate']
+                if config_data['draw_shape']['shape'] == 'Line':
+                    can.line(
+                        x1=x0 * inch,
+                        y1=y0 * inch,
+                        x2=x1 * inch,
+                        y2=y1 * inch
+                    )
+                elif config_data['draw_shape']['shape'] == 'Rectangle':
                     fill = config_data['draw_shape'].get('fill', 1)
+                    can.rect(
+                        x=x0 * inch,
+                        y=y0 * inch,
+                        width=x1 * inch,
+                        height=y1 * inch,
+                        fill=fill
+                    )
 
-                    can.rect(x0*inch,y0*inch, x1*inch, y1*inch, fill=fill)
             elif 'image' in config_data:
-                x = config_data['image']['x-coordinate']
-                y = config_data['image']['y-coordinate']
-                w = config_data['image']['width']
-                h = config_data['image']['height']
-                can.drawImage(values[key], x, y, width=w, height=h)
+                can.drawImage(
+                    image=self.values[key],
+                    x=config_data['image']['x-coordinate'],
+                    y=config_data['image']['y-coordinate'],
+                    width=config_data['image']['width'],
+                    height=config_data['image']['height']
+                )
 
             else:
-                x = config_data['x-coordinate']
-                y = config_data['y-coordinate']
-                value = config_data.get('value', None)
-
-                if not value:
-                    value = values[key]
-
-                try:
-                    can.drawString(x, y, value)
-                except NameError:
-                    raise ValueError('could not find co ordinates for key({}) value {}'.format(key, values[key]))
+                can.drawString(
+                    x=config_data['x-coordinate'],
+                    y=config_data['y-coordinate'],
+                    text=config_data.get('value', self.values[key])
```

Comment:
@joeirimpan it breaks here 

config_data.get('value') is the default value so some times key wont exists but it will break here since you are checking for key(self.values['key])

Reply:
@shridarpatil The key `value` inside config_data is not guaranteed right? The fallback is self.values[key] which i m assuming will always be there which otherwise raises keyerror. There should be an assert statement which guarantees self.values[key] since we are freely using it everywhere
https://github.com/zerodhatech/pdf_text_overlay/blob/master/pdf_text_overlay/pdfWriter.py#L65
https://github.com/zerodhatech/pdf_text_overlay/blob/master/pdf_text_overlay/pdfWriter.py#L71
https://github.com/zerodhatech/pdf_text_overlay/blob/master/pdf_text_overlay/pdfWriter.py#L123